OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -110,7 +113,20 @@ public bool CanEnterCell(World world, Actor self, CPos cell, SubCell subCell = S
 			if (locomotor.MovementCostForCell(cell) == short.MaxValue)
 				return false;
 
-			return locomotor.CanMoveFreelyInto(self, cell, subCell, check, ignoreActor);
+			if (!locomotor.CanMoveFreelyInto(self, cell, subCell, check, ignoreActor))
+				return false;
+
+			// We cannot cross diagonals if another actor is already moving diagonal.
+			if (!CheckAdjacentOnDiagonal || !fromCell.HasValue)
+				return true;
+
+			var direction = cell - fromCell.Value;
```

Comment:
Shouldn't all this be living inside `locomotor.CanMoveFreelyInto`?

It doesn't seem right that `locomotor.MovementCostToEnterCell` internally accounts for `CheckAdjacentOnDiagonal` but that the callers must manage this externally for `CanMoveFreelyInto`.